components/plan-build/bin/environment.bash

```diff
@@ -204,80 +212,61 @@ __populate_environment_from_deps() {
-# and utterly paranoid, we'll iterate through the TDEPS of the
-# dependency and assemble a PATH by layering all the PATHs of each of
-# those dependencies.
+# **Internal**  Build a `PATH` string suitable for entering into the `PATH` key
+# of this package's `RUNTIME_ENVIRONMENT` metadata file. The ordering of this
+# path is important as this value will ultimately be consumed by other programs
+# such as the Supervisor when constructing the `PATH` environment variable
+# before spawning a process.
 #
-# Don't need to do this for packages that already have a
-# RUNTIME_ENVIRONMENT file, because this operation will have already
-# been done when those packages were built. This is to help preserve
-# backward compatibility.
-__assemble_legacy_path() {
-    local path_to_dep="${1}"
-    local path=""
-
-    local tdeps
-    tdeps=$(_read_metadata_file_for ${path_to_dep} "TDEPS")
-    if [[ -n "${tdeps}" ]]; then
-        for tdep in ${tdeps}; do
-            local path_to_tdep="${HAB_PKG_PATH}/${tdep}"
-            # If a PATH exists for this transitive dependency, collect
-            # it, prepending to the beginning of our accumulating path
-            local tdep_path
-            tdep_path=$(_read_metadata_file_for "${path_to_tdep}" "PATH")
-            if [[ -n "${tdep_path}" ]]; then
-                path=$(push_to_path "${tdep_path}" "${path}")
-            fi
-        done
-    fi
-    # If the current dependency has PATH data itself, prepend it to
-    # what we've been accumulating so far
-    local current_dep_path
-    current_dep_path=$(_read_metadata_file_for "${path_to_dep}" "PATH")
-    if [[ -n "${current_dep_path}" ]]; then
-        path=$(push_to_path "${current_dep_path}" "${path}")
+# The path is constructed by taking all `PATH` metadata file entries from this
+# package (in for the form of `$pkg_bin_dirs[@]`), followed by entries from the
+# *direct* dependencies first (in declared order), and then from any remaining
+# transitive dependencies last (in lexically sorted order). All entries are
+# present only once in the order of their first appearance.
+__assemble_runtime_path() {
+  local paths=()
+  local dir dep data
+
+  # Add element for each entry in `$pkg_bin_dirs[@]` first
+  for dir in "${pkg_bin_dirs[@]}"; do
+    paths+=("$pkg_prefix/$dir")
+  done
+
+  # Iterate through all direct direct run dependencies following by all
+  # remaining transitive run dependencies and for each, append each path entry
+  # onto the result, assuming it hasn't already been added. In this way, all
+  # direct dependencies will match first and any programs that are used by a
+  # direct dependency will also be present on PATH, albeit at the very end of
+  # the PATH.
+  for dep in "${pkg_deps_resolved[@]}" "${pkg_tdeps_resolved[@]}"; do
+    if [[ -f "$dep/PATH" ]]; then
+      data="$(cat "$dep/PATH")"
+      data="$(trim "$data")"
+      while read -r entry; do
+        paths=($(_return_or_append_to_set "$entry" "${paths[@]}"))
```

Comment:
We might want to filter out any elements in the `PATH` file that aren't rooted in the package it's coming from, to account for packages built when `PATH` contained the entire runtime path of the package.

Reply:
That's a great, excellent point. Turns out this was easy and a regression of my own making here, so I'll rebase this into the commit and augment the comment above this code block to explain.

Reply:
I also renamed `$dep` to `$dep_prefix` in order to better explain that this variable contains a prefix path and not a package identifier, which in turn makes the above suggestion clearer to implement.